File: src/test/java/accounting/patterns/AnnualServiceFeeRule.java
package accounting.patterns;

import java.math.BigDecimal;
import java.util.Currency;

/**
 * A flat annual fee, charged per unit of quantity and independent of the
 * service agreement's rate.
 */
record AnnualServiceFeeRule(EntryType entryType, long amount) implements PostingRule {
    private static final Currency USD = Currency.getInstance("USD");

    @Override
    public MonetaryAmount calculateAmount(Quantity quantity, BigDecimal rate) {
        return MonetaryAmount.of(USD,
                BigDecimal.valueOf(quantity.value()).multiply(BigDecimal.valueOf(this.amount)));
    }
}

File: src/test/typescript/accounting-pattern/AnnualServiceFeeRule.ts
import { PostingRule } from "../../../main/typescript/accounting-pattern/PostingRule";
import { EntryType } from "../../../main/typescript/accounting-pattern/EntryType";
import { Quantity } from "../../../main/typescript/accounting-pattern/Quantity";
import { MonetaryAmount } from "../../../main/typescript/accounting-pattern/MonetaryAmount";

class AnnualServiceFeeRule extends PostingRule {
    private amount: number;

    constructor(service: EntryType, amount: number) {
        super(service);
        this.amount = amount;
    }

    protected calculateAmount(quantity: Quantity, rate: number): MonetaryAmount {
        return new MonetaryAmount("USD", quantity.getValue() * this.amount);
    }
}
